asmcomp/CSEgen.ml

```diff
@@ -227,8 +227,14 @@ method class_of_operation op =
   | Iextcall _ | Iopaque -> assert false       (* treated specially *)
   | Istackoffset _ -> Op_other
   | Iload { mutability; is_atomic } ->
-      (* #12173: disable CSE for atomic loads. *)
-      if is_atomic then Op_other
+      (* #12173: disable CSE for atomic loads.
+         #12825: atomic loads cannot be treated as Op_other
+           because they update our view / the frontier of the
+           non-atomic locations, so past non-atomic (mutable) loads
+           may be not be valid anymore.
+         We conservatively tread them as non-initializing stores.
+      *)
+      if is_atomic then Op_store true
```

Comment:
Cosmetic change suggestion.

Is it worth changing the definition of `Op_load` to include atomicity information (similar to`Iload {is_atomic: bool; mutability: Asttypes.mutable_flag`)? `Op_load {is_atomic = true}` will be treated the same as `Op_store true`.

https://github.com/ocaml/ocaml/blob/ba71f2e47d3c260196bbdb9c89b24517e6c2fab8/asmcomp/CSEgen.ml#L25-L30

Reply:
We could do this as a refactor that also changes `Op_store of bool` into something like `Op_store of [ Init | Assign ]`. I would suggest to defer this to a different PR to give more time to the questions of taste than to the bugfix.

Reply:
Sounds good.